File: TinyScript/JAVA/src/parser/ast/Factor.java
package parser.ast;

import lexer.TokenType;
import parser.util.PeekTokenIterator;

// 因子（等号两边那些东西）类，可以用来运算等
public abstract class Factor extends ASTNode {
    public Factor(ASTNode _parent, PeekTokenIterator it) {
        super(_parent);
        var token = it.next();
        var type = token.getType();

        // 凭借下个流数据判断类型
        if (type == TokenType.VARIABLE) {
            this.type = ASTNodeTypes.VARIABLE;
        } else {
            this.type = ASTNodeTypes.SCALAR;
        }

        this.label = token.getValue();
        this.lexeme = token;
    }
}

File: TinyScript/JS/src/parser/ast/Factor.js
const ASTNode = require("./ASTNode");
const TokenType = require('../../lexer/TokenType')
const ASTNodeTypes = require('../ast/ASTNodeTypes')

class Factor extends ASTNode {
  constructor(parent, it) {
    super(parent)
    const token = it.next()

    var type = token.getType()

    // 在 初始化Factor 时就给一个流数据赋类型值
    if (type === TokenType.VARIABLE) {
      this.type = ASTNodeTypes.VARIABLE
    } else {
      this.type = ASTNodeTypes.SCALAR
    }

    this.label = token.getValue()
    this.lexeme = token
  }
}

module.exports = Factor